### ISSUE TITLE
Set up Recipe entities and database

### ISSUE THREAD

**Issue by BenasB** [BenasB/vu-psk#8]
It's probably best to use [EF Core](https://learn.microsoft.com/en-us/ef/core/) for data access (If you feel like it, you could use something else, e.g. Dapper).

- Set up the `Recipe` entity and the `DbContext`
  - Think about the needed fields (feel free to discuss/introduce ideas). The entity will need to be mapped to [this DTO](https://github.com/BenasB/vu-psk/blob/ecfc2e5cf9c3553349539a5eb1eea7e7fdefe10a/Recipes/Recipes.Public/Recipe.cs) (of course this can be adjusted) later down the line.
  - An `Ingredient` should also probably be an entity, think of the relationship between recipe and ingredient (taking into account that a recipe has, e.g. "500g" of "Sugar" ingredient). 
- Choose and set up a SQL database and make an initial migration.
  - Ideally there should be a repeatable way of the DB setup. Maybe a [docker compose](https://docs.docker.com/compose/) file at the repo's root folder that would include all of the development dependencies? Or at least some instructions/docs around that.

**Comment by adomas-vensas** [comment]
![Image](https://github.com/BenasB/vu-psk/assets/90962230/f450b3e8-0241-4edf-ba3f-a45d3a676ca6)

This is an initial sketch of the entities concerning Recipes. I searched for some entity ideas on `beatosvirtuve.lt` and `allrecipes.com`

- Recipes:
  - I think all fields are self-explanatory. `author` and `author_id` fields are separate so that only the author's name can be loaded without the need to load the entire User entity.
- Nutrition (one-to-one with Recipes):
  - Nutrition table holds information about recipe's nutritional value per serving. `calories` can be calculated from these fields and should be a separate fields in the entity.
- Instructions (many-to-one with Recipes):
  - As seen on different recipe websites, all of them list instructions as an ordered list of strings. This can be achieved by the `step` field that should self-increment and create ordering within instructions.
- Tags (many-to-many with Recipes):
  - There can be a variety of tags not necessarily only for food categories but for "Quick Recipes", "Cheap", etc., so tags are kept as simple strings. Tags also have an `amount` field that tracks how many times they have been used.
- Ingredients (many-to-one with Recipes):
  - As seen on different recipe websites, they simply list the ingredients without considering what units or foods are actually listed in them. If we wanted to consider the units and contents of the ingredient (e.g., "500g of sugar"), it's probably best to check this in the front-end when creating the recipe.

Ideas:
-  Should we consider adding a field for recipe photos right away?

(I don't know if everyone will see this so I'm @'ing everyone 😄 @BenasB @Rytis-Sapka @6ar8nas @TomasKvaraciejus)

**Comment by Rytis-Sapka** [comment]
My insights:
- Postgre SQL supports arrays so the ```Instructions``` can be made into a string array inside ```Recipes```
- ```Amount``` field in ```Tags``` could just be achieved by a query and probs shouldn't be saved
- If we are using ```AuthorId``` then we should also probaby have an ```Author``` table (or maybe this will be handled by the other API?)
- I think it would be nice to make ```Ingredients``` and ```Recipies``` many-to-many. This could enable easier filtering in the future if we wanted to do that.
- It won't hurt to add ```quantity``` to ```Ingredients```. We can just ignore it in the UI.

**Comment by TomasKvaraciejus** [comment]
AuthorId could be a FK to the Users table which will need to created at some point anyway. There is a separate issue concerning this (#14 but GitHub won't let me link it for some reason).

**Comment by 6ar8nas** [comment]
Regarding the image - I think we can use a placeholder for now and include it at some later point.

I'd advocate for dropping the "amount" property on Tag entity. I also think "author" property on Recipe is not needed.

It's a bit confusing to me why ingredients and recipes are a M-1 and not M-M. It would seem that one recipe can have many ingredients, but one ingredient may also be used in many recipes.

The units of measurement part could be achieved by having an intermediary table RecipeIngredients. It could have a recipe_id, ingredient_id, units of measurement and the amount. In such case, nutrition could also fall under Ingredient and that could also be accumulated through some multiplication over units of measurement instead of entered by the user.

If we go to the detail of nutrition, we could also add allergens to ingredients, so that our recipes would also include accumulated information about it, but that's just an option for how we could perhaps expand our application in the future.

**Comment by adomas-vensas** [comment]
All right, I will remove the `amount` on the Tags table and `author` from recipes. They do seem redundant.

Yes, I think we should follow the standardized format of instructions going in order. Authors of recipes themselves could direct the cooks to do some instructions in parallel if they wanted.

About the ingredient part. This is how `allrecipes.com` manages the addition of ingredients. All of them are added as strings of information, even quantities.
![image](https://github.com/BenasB/vu-psk/assets/90962230/d5257f07-375c-4bfa-b1f2-f0e76ad28eff)

From this, I suppose that ingredients are considered unique for every recipe regardless of if they repeat or not, hence the M-1 relationship.

Of course, if we want to, I can create the suggested structure for the `Ingredients` table, where nutritional value is measured for 100g and then convert that information to different measurements without the manual input from users. Here's a quick sketch:
![image](https://github.com/BenasB/vu-psk/assets/90962230/09046bfe-3405-4348-b277-676760815368)

Yes, adding allergens would be a nice touch. We could add this as a secondary objective for the project.

**Comment by 6ar8nas** [comment]
My only issue with the allrecipes.com approach is the fact that we could probably no longer filter by ingredients as easily, though I'm unsure whether that's a priority. The relationship makes more sense now though and it might be simpler too.

**Comment by adomas-vensas** [comment]
All good, I fully understand. The `allrecipes` approach is simpler but we than should decide if we really want to filter by ingredients. `Allrecipes` approach makes it probably impossible to do it efficiently. Maybe filtering by `tags`, `serving_size` and `cooking_time` is enough?

**Comment by 6ar8nas** [comment]
I guess. Given that we also have the search feature in our plan, that might be an alternative to work around this need. Also if that's not a problem we could wait for @BenasB input.

**Comment by BenasB** [comment]
Okay, so the big thing (as alluded in the task) is "do we treat ingredients as standalone entities?"

I see you've boiled it down to the main thing: "do we want to filter on ingredients or not". If we consider the ingredients as standalone entities, then we could have a list of all ingredients in the system and effectively search through them. If we don't consider the ingredients as standalone entities, then we can probably do some sort of fuzzy search through them.

I see we are more leaning toward the "ingredients are part of a recipe" (not standalone) way and that's fine, since it makes things less complicated. I don't think there are any requirements of keeping track of ingredients. BTW the design I did in Figma assumed we were trading the ingredients as standalone, but that can just simply be adjusted.

I believe that these things:
- Nutrition
- Allergies
- Even maybe tags
Can wait a bit and be implemented with a separate task, since this one should focus on getting something basic working

**Comment by adomas-vensas** [comment]
Okay, so if we go forth with the "ingredients are part of the recipe" solution, this might be the first `Recipes` Entity. I removed the `Tags` and `Nutrition` tables as recommended by @BenasB. Also per recommendation of @Rytis-Sapka I also removed `Ingredients` and `Instructions` entities and left them as lists.

Does this structure seem sufficient?

![Image](https://github.com/BenasB/vu-psk/assets/90962230/6a54491f-e1d6-4d3f-897b-0d629144b40b)

**Comment by BenasB** [comment]
I am not familiar with how lists work in PostgreSQL, but I assume that's just a more DB managed one to many relationship.
Will EF Core understand how to use those types? If so, all is well

**Comment by adomas-vensas** [comment]
Maybe @Rytis-Sapka could comment on this

**Comment by BenasB** [comment]
Cool, sounds fine then :D
